godot-core/src/private.rs

```diff
@@ -261,19 +280,84 @@ pub(crate) fn has_error_print_level(level: u8) -> bool {
     ERROR_PRINT_LEVEL.load(atomic::Ordering::Relaxed) >= level
 }
 
+#[cfg(debug_assertions)]
+struct ScopedFunctionStack {
+    functions: Vec<*const dyn Fn() -> String>,
+}
+
+/// Internal type used to store context information for debug purposes. Debug context is stored on the thread-local
+/// ERROR_CONTEXT_STACK, which can later be used to retrieve the current context in the event of a panic. This value
+/// probably shouldn't be used directly; use ['get_gdext_panic_context()'](get_gdext_panic_context) instead.
+#[cfg(debug_assertions)]
+impl ScopedFunctionStack {
+    /// # Safety
+    /// Function must be removed (using [`pop_function()`](Self::pop_function)) before lifetime is invalidated.
+    unsafe fn push_function(&mut self, function: &dyn Fn() -> String) {
+        /// # Safety
+        /// The caller must ensure that the function isn't used past its original lifetime.
+        /// (Invariant must be held by push_function)
+        unsafe fn assume_static_lifetime(
+            value: &dyn Fn() -> String,
+        ) -> &'static dyn Fn() -> String {
+            std::mem::transmute(value)
+        }
+        self.functions
+            .push(assume_static_lifetime(function) as *const _);
+    }
+
+    fn pop_function(&mut self) {
+        self.functions.pop().expect("function stack is empty!");
+    }
+
+    fn get_last(&self) -> Option<String> {
+        self.functions.last().cloned().map(|pointer| {
+            // SAFETY:
+            // Invariants provided by push_function assert that any and all functions held by ScopedFunctionStack
+            // are removed before they are invalidated; functions must always be valid.
+            unsafe {
+                &*pointer
+            }
+        }())
+    }
+}
+
+#[cfg(debug_assertions)]
+thread_local! {
+    static ERROR_CONTEXT_STACK: RefCell<ScopedFunctionStack> = const {
+        RefCell::new(ScopedFunctionStack { functions: Vec::new() })
+    }
+}
+
+// Value may return `None` even from panic hook if called from a non-Godot thread
```

Comment:
Also here, you didn't apply my suggestion verbatim, yet resolved the conversation.
https://github.com/godot-rust/gdext/pull/1037#discussion_r1948035569